compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrUtils.kt

```diff
@@ -212,20 +212,31 @@ private fun IrTypeParameter.copySuperTypesFrom(source: IrTypeParameter) {
     }
 }
 
+// Copy value parameters, dispatch receiver, and extension receiver from source to value parameters of this function.
+// Type of dispatch receiver defaults to source's dispatch receiver. It is overridable in case the new function and the old one are used in
+// different contexts and expect different type of dispatch receivers. The overriding type should be assign compatible to the old type.
 fun IrFunction.copyValueParametersToStatic(
     source: IrFunction,
-    origin: IrDeclarationOrigin
+    origin: IrDeclarationOrigin,
+    dispatchReceiverType: IrType? = null
```

Comment:
Why not `source.dispatchReceiverParameter?.type`?

Reply:
Assuming `C.foo()` is a fake override of `P.foo()`. `source.dispatchReceiverParameter?.type` is `P`. In `C.access@foo()`, passing it as type `P` to `C.foo()` would upset the bytecode verifier, because `P` isn't assign compatible to `C`.

Other approaches considered are:
1. Always use source.parent.type. Although all the tests passed, I'm not sure whether source.parent is always the desired class.
2. Insert a runtime cast. There are overheads, though.
3. In the accessor, change the call of the fake override to the real function. This is simplest, but it makes a few other errors. Some of them seem to require usage of `KotlinType`, so I didn't investigate further.

Reply:
My concern was not so much the program logic, but clarity of the code. Here, passing `null` invokes a special behaviour, and it is not clear from the function header what that behaviour is.

Reply:
Thanks for the clarification. I've updated it accordingly. Please take a look again.